src/client_builder.rs

```diff
@@ -61,38 +134,91 @@ where
         ClientBuilder {
             domain,
             port,
-            starttls: false,
+            mode: ConnectionMode::Tcp,
+            #[cfg(any(feature = "native-tls", feature = "rustls-tls"))]
+            tls_kind: TlsKind::Any,
+            #[cfg(any(feature = "native-tls", feature = "rustls-tls"))]
+            no_tls_verify: false,
         }
     }
 
     /// Use [`STARTTLS`](https://tools.ietf.org/html/rfc2595) for this connection.
-    #[cfg(any(feature = "native-tls", feature = "rustls-tls"))]
     pub fn starttls(&mut self) -> &mut Self {
-        self.starttls = true;
+        self.mode = ConnectionMode::StartTls;
         self
     }
 
-    /// Return a new [`Client`] using a `native-tls` transport.
-    #[cfg(feature = "native-tls")]
-    #[cfg_attr(docsrs, doc(cfg(feature = "native-tls")))]
-    pub fn native_tls(&mut self) -> Result<Client<TlsStream<TcpStream>>> {
-        self.connect(|domain, tcp| {
-            let ssl_conn = TlsConnector::builder().build()?;
-            Ok(TlsConnector::connect(&ssl_conn, domain, tcp)?)
-        })
+    /// Use TLS for this connection.
+    #[cfg(any(feature = "native-tls", feature = "rustls-tls"))]
+    pub fn tls(&mut self) -> &mut Self {
+        self.mode = ConnectionMode::Tls;
+        self
     }
 
-    /// Return a new [`Client`] using `rustls` transport.
-    #[cfg(feature = "rustls-tls")]
-    #[cfg_attr(docsrs, doc(cfg(feature = "rustls-tls")))]
-    pub fn rustls(&mut self) -> Result<Client<RustlsStream<TcpStream>>> {
-        self.connect(|domain, tcp| {
-            let ssl_conn = RustlsConnector::new_with_native_certs()?;
-            Ok(ssl_conn.connect(domain, tcp)?)
-        })
+    /// Sets the TLS backend to use for this connection.
+    #[cfg(any(feature = "native-tls", feature = "rustls-tls"))]
+    pub fn tls_kind(&mut self, kind: TlsKind) -> &mut Self {
```

Comment:
I wonder if this should auto-set `self.mode = Tls` if it's currently set to `Tcp`. At least, it feels weird to have to call `tls_kind` and _also_ `tls`.

Reply:
Well, hopefully any "reasonable" IMAP server is properly configured to disallow non-TLS connections in that case :)  ( I know my public facing IMAP servers are configured this way)

So a few things, 
1) if a user did connect to port 993 without calling TLS() it wouldn't connect.
2) The issue with defaulting TLS on is you are assuming they are connecting to port 993?   (thus port 143 wouldn't work).
3) Should we perhaps maybe introspect the port # passed in and default to TLS if 993 is passed? and starttls if port 143 is passed?
4) Or maybe have it default to an "auto mode" where by
   a) if port = 993 then TLS on
   b) otherwise use starttls *IF* the server supports it (e.g. we would introspect the initially advertised capabilities in the greeting, which a server support STARTTLS should be advertising there)

Honestly #4 would provide the nicest use-case as it would be security by default (if possible). and provide the ease of use you are wanting to give in the crate API.  And uses who want to "go beyond" can simply call the explicit `tls()`, `starttls()`, and (new) `raw()` methods to change the connection mode explicitly.


For "connect_with", we'd still have to figure that out.  Maybe changing the contract to something like this.
`connect_with(starttls: bool, handshake: FnOnce)`
Thus we have to explicitly pass whether we are expecting to do starttls or not.  (false if we are doing RAW or full TLS). Thus allowing that the full flexibility, and it would essentially "ignore" the connection mode enum in the connection object.

Reply:
I like doing number 4!

For `connect_with`, I guess I wonder why we'd want to have a separate `bool` that overrides whatever `starttls` is set to, rather than just respecting what it's set to directly?

Reply:
So, I have "auto" mode implemented locally for the main builder.

Now onto connect_with

So, my thinking of why connect_with would take the starttls param is that it's a "manual" connection mode anyway thus it isn't using any of the "knowledge" from the builder other than the port and host.  (e.g. its even ignoring "tls_kind").

if I have it rely on the ConnectionMode instead then the question is what happens in each scenario.

1. Auto
    - if 993 then TLS (leave alone) otherwise connect and "detect" starttls. Though we really need to "tell" the FnOnce what it is receiving in this case (raw or TLS connection)
2. Tcp/Raw
    - just pass the connection through to the callback
3. StartTls
    - upgrade the TLS always with starttls and pass the upgraded connection to the callback
4. Tls
    - just pass the connection through the callback

Thus to make this really work cleanly we'd really need to change the contract of the callback to accept the socket and the mode so that it knows if it has a Tcp/Raw, or TLS-ready connection has been passed.  or do we ONLY call the callback when we have a TLS connnection?

This goes back to what are the use-cases of this method?

Is it to provide TLS customization? (certs, etc..). If so we can provide better methods to handle that (as I explained elsewhere).

Is it to provide full customization of the TCP connection in other ways?  If so, do we really need to provide that in the ClientBuilder and can't the developer just create a TCP connection themselves?

Is the handling of StartTls in this method making it more difficult?

Anyways,

I'm pushing up a commit that

1) adds an "auto"mode as discussed in point 4 above
2) reworks the connect_with to utilize the "Auto" mode.

I'm going to play around a little more later tonight and possible rewrite the `connect` method to use connect_with and make the code smaller.

Reply:
I think I'm sold, including probably just removing this kind of "provide a handler" method. As you say, for customizing TCP users can just do that before passing the connection to `Client`, and for TLS configuration we can do better anyway.

Part of what sold me here is also that the current release version of `imap` doesn't even have this kind of a method, so it can't be something a lot of people depend on :p